backend/src/main/java/fk/prof/backend/model/association/impl/ZookeeperBasedBackendAssociationStore.java

```diff
@@ -0,0 +1,358 @@
+                  /**
+                   * Presently assigned backend is defunct, find an available backend and if found, de-associate current backend and assign the new one to process group
+                   * Defunct backend is not de-associated eagerly because if no available backend is found, its better to wait for current backend to come back alive
+                   */
+                  availableBackendsByPriority.remove(existingBackend);
+                  BackendDetail newBackend = getAvailableBackendFromPriorityQueue();
+                  try {
+                    if (newBackend == null) {
+                      logger.warn(String.format("Presently assigned backend=%s for process_group=%s is defunct but cannot find any available backend so keeping assignment unchanged",
+                          existingBackendAssociation, ProtoUtil.processGroupCompactRepr(processGroup)));
+                      future.complete(existingBackendAssociation);
+                    } else {
+                      try {
+                        deAssociateBackendWithProcessGroup(existingBackend, processGroup);
+                        associateBackendWithProcessGroup(newBackend, processGroup);
+                        if (logger.isDebugEnabled()) {
+                          logger.debug(String.format("process_group=%s, de-associating existing backend=%s, associating new backend=%s",
+                              ProtoUtil.processGroupCompactRepr(processGroup), existingBackend.getBackendIPAddress(), newBackend.getBackendIPAddress()));
+                        }
+                        future.complete(newBackend.getBackendIPAddress());
+                      } catch (Exception ex) {
+                        future.fail(String.format("Cannot persist association of backend=%s with process_group=%s in zookeeper",
+                            newBackend.getBackendIPAddress(), processGroup));
+                      }
+                    }
+                  } finally {
+                    if (newBackend != null) {
+                      try {
+                        availableBackendsByPriority.offer(newBackend);
+                      } catch (Exception ex) {
+                        //TODO: Some metric to indicate enqueue failure here
+                        logger.error("Error enqueuing backend=" + newBackend.getBackendIPAddress() + " back in the queue");
+                      }
+                    }
+                  }
+                }
+              }
+            } finally {
+              backendAssignmentLock.unlock();
+            }
+          } else {
+            future.fail("Timeout while acquiring lock on backend queue for process_group=" + processGroup);
+          }
+        } catch (InterruptedException ex) {
+          future.fail(new RuntimeException("Interrupted while acquiring lock on backend queue for process_group=" + processGroup, ex));
+        }
+      }, result.completer());
+    }
+
+    return result;
```

Comment:
The response-reordering race is possible with association too, but recorder doesn't issue multiple requests at a time, so as long as backend doesn't snoop the response (and cache it or something of that nature), this is safe.